arch/risc-v/src/mpfs/mpfs_dsn.c

```diff
@@ -0,0 +1,154 @@
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <errno.h>
+#include <unistd.h>
+#include "mpfs_dsn.h"
+#include "riscv_internal.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MPFS_SCBCTRL_BASE  0x37020000ul
+#define SERVICES_CR_OFFSET 0x50
+#define SERVICES_SR_OFFSET 0x54
+
+#define SERVICES_CR        (MPFS_SCBCTRL_BASE + SERVICES_CR_OFFSET)
+#define SERVICES_SR        (MPFS_SCBCTRL_BASE + SERVICES_SR_OFFSET)
+
+/* Command bits */
+
+#define SCBCTRL_SERVICESCR_REQ  (1 << 0)
+
+/* Status bits */
+
+#define SCBCTRL_SERVICESSR_BUSY (1 << 1)
+
+/* 2kB long mailbox. */
+
+#define MSS_SCBMAILBOX     0x37020800ul
+
+/* Retry count */
+
+#define RETRIES            500
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_read_dsn
+ *
+ * Description:
+ *   Read n bytes of the device serial number. Full serial number is 16 bytes
+ *
+ * Parameters:
+ *   dsn - A pointer to the destination buffer
+ *   len - Number of bytes to read
+ *
+ * Returned Value:
+ *   Number of bytes read, -ETIMEDOUT on error
+ *
+ ****************************************************************************/
+
+int mpfs_read_dsn(uint8_t *dsn, size_t len)
+{
+  uint32_t reg;
+  uint8_t *p = (uint8_t *)MSS_SCBMAILBOX;
+  irqstate_t flags = enter_critical_section();
+  unsigned retries = RETRIES;
+
+  /* Wait until the system controller is not busy.
+   * Read the SN inside critical section, just in case someone else is
+   * using the system controller services
+   */
+
+  while ((getreg32(SERVICES_SR) & SCBCTRL_SERVICESSR_BUSY) && --retries > 0)
+    {
+      leave_critical_section(flags);
+      usleep(1000);
+      flags = enter_critical_section();
```

Comment:
I need to double check, but I think it should be fine jus to
```suggestion
      usleep(1000);
```

Reply:
Yes, usleep will leave and enter the critical section automatically.

Reply:
Agree, what is done here is silly. You can just as well enter the critical section before the while loop. It will schedule away at usleep anyhow. Also, it is better to go to sleep in other while loops as well during re-tries in order not to block the interrupts for a long time. I will change this to more "sane".

Reply:
No, reverted this to original.

The intention of the first loop is NOT to block all interrupts for a calling thread for the whole time of waiting for the HW block to become free.

The intention is to continue the function in critical section after the HW block is free.